Reproduce Guinness example With OSRM docker image

In [1]:
import geopandas as gpd
import pandas, numpy, pyvrp, sys

In [2]:
sys.path.insert(0, '/home/dylan/projects/gsoc2025/spopt/') # not the published spopt

In [3]:
import spopt
print(spopt.__file__)

/home/dylan/projects/gsoc2025/spopt/spopt/__init__.py


In [4]:
from spopt.route import engine, heuristic, utils

## Reproduce the guinness example

In [5]:
from spopt.route.heuristic import LastMile
from pyvrp import stop

In [6]:
trucks = pandas.DataFrame(
    [['big', 'lng',      2000,    280, .004,  .50, 5],
     ['big', 'electric', 2000,    480, .002,  .50, 5],
     ['med', 'lng',      800, 280*.66, .0001, .63, 10],
     ['med', 'electric', 800, 480*.66, .004,  .50, 10],
     ['smo', 'lng',      400, 280*0.4, .002,  .50, 20],
     ['smo', 'electric', 400, 480*0.4, .0001, .63, 20],
     ],
     columns = [
         'namesize', 'namefuel', 'capacity', 
         'fixed_cost', 'cost_per_meter', 'cost_per_minute', 'n_truck'
         ]
)

In [7]:
dublin_pubs = gpd.read_file('/home/dylan/projects/gsoc2025/spopt/notebooks/gsoc2025/data/dublinpubs.geojson')

In [8]:
dublin_pubs.shape

(551, 8)

In [9]:
gdf = dublin_pubs

In [10]:
depot = gdf.iloc[0,:]
clients = gdf.iloc[1:,:].reset_index(drop=True)
clients = clients.set_index(clients.osmid.astype(str))

In [11]:
_clients = clients[:90]

I am running into rate limits using the osrm docker image. When using the full set of clients (e.g. all 551 pubs), the returned json content from osrm (engine.py, line 163 when trying to parse `content`) is:

```
{'code': 'TooBig', 'message': 'Too many table coordinates'}
```

Solution: add new parameter for the docker service call:

``
—max-table-size 9999999
``

In [12]:
print('initializing model')
m = LastMile(
    depot_location=(depot.longitude.item(), depot.latitude.item()),
    depot_open=pandas.Timestamp("2030-01-02 07:00:00"),
    depot_close=pandas.Timestamp("2030-01-02 20:00:00"),
    depot_name=depot['name'],
)
print("adding clients")

m.add_clients(
    locations = clients.geometry, 
    delivery = clients.demand,
    pickup = clients.supply,
    time_windows=None,
    service_times=(numpy.log(clients.demand)**2).astype(int)
)
print("adding trucks")
m.add_trucks_from_frame(
    trucks, 
)

initializing model
adding clients
adding trucks


In [13]:
m.depot_location

(-6.28688, 53.341972)

In [14]:
m.solve(stop=pyvrp.stop.MaxRuntime(300))

PyVRP v0.11.3

Solving an instance with:
    1 depot
    550 clients
    70 vehicles (6 vehicle types)

                  |       Feasible        |      Infeasible
    Iters    Time |   #      Avg     Best |   #      Avg     Best
H     500    188s |  42 49057499 39473545 |  44 36599331 35408684

Search terminated in 300.00s after 823 iterations.
Best-found solution has cost 39171252.

Solution results
    # routes: 12
     # trips: 12
   # clients: 550
   objective: 39171252
    distance: 285939
    duration: 1503
# iterations: 823
    run-time: 300.00 seconds



In [15]:
m.write_result("docker-image")